compiler/rustc_parse/src/parser/pat.rs

```diff
@@ -127,6 +107,62 @@ impl<'a> Parser<'a> {
         Ok(self.mk_pat(or_pattern_span, PatKind::Or(pats)))
     }
 
+    /// Parse the pattern for a function or function pointer parameter.
+    pub(super) fn parse_fn_param_pat(&mut self) -> PResult<'a, P<Pat>> {
+        // We actually do _not_ allow top-level or-patterns in function params, but we use
```

Comment:
Why don't we allow them?

Reply:
My understanding from the or-patterns tracking issue and the team consensus was that closures should not support them, and:

> This also applies to functions which, although do not look as ambiguous, benefit from better consistency with closures. With respect to function arguments there's also the issue that not disambiguating with parenthesis makes it less clear whether the type ascription applies to the or-pattern as a whole or just the last alternative.

Reply:
I see, consistency with closures.

For addressing the type ascription issue it would be enough to keep them unstable, similarly to `let a | b: u8`.
Actually, `let a | b: u8` could be turned into an error too as an alternative to unstability.

Reply:
Sounds good, but I think I will make this change in a followup PR if you don't mind.